# Building a quick and dirty bigram LM

First import some useful libraries.

In [1]:
from collections import Counter

In [2]:
from collections import defaultdict

In [52]:
import numpy as np

Next let's pull in all the data into one long string (note this is not really a great idea from a practical programming perspective).

In [182]:
text = open("hobbit-train.txt").read()

In [183]:
len(text)

521691

Split on white space. 'words' is now just a list of strings.

In [184]:
words = text.split()

In [174]:
len(words)

95940

Let's use a Counter to get the frequencies of the words in the corpus. A Counter is basically a dictionary that keeps track of the frequency of items in a collection. The keys are the items and the values are their frequencies in the collection.

In [185]:
rawfreqs = Counter(words)

In [186]:
len(rawfreqs)

12333

So there are 12333 unique strings in the Hobbit (using whitespace to split on). One of them is surely 'Gandalf'.

In [187]:
rawfreqs["Gandalf"]

99

Keeping track of words and their counts using strings isn't very efficient. So before moving on we'll map known words to unique integer indexes. To do this we'll use the defaultdict flavor of dictionary that takes a method that specifies how to create values for new entries to the index. So in the following, word2index is a dictionary where the keys are strings and the values are the integer index for each string. We'll assign the index based on when the word is encountered as we walk through the text.

In [188]:
word2index = defaultdict(lambda: len(word2index))

And let's make <UNK> the first entry to the index.

In [189]:
UNK = word2index["<UNK>"]

In [190]:
word2index["<UNK>"]

0

In [191]:
UNK

0

Now let's build our actual lexicon. We'll use the heuristic that we can ignore words that only happened once in training.

In [192]:
[word2index[word] for word, freq in rawfreqs.items() if freq > 1]

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,
 185

Do that we're down to roughly 5000 words from 12000 in our vocabulary. 

In [193]:
len(word2index)

5209

In [194]:
len(rawfreqs)

12333

Now we need to freeze the lexicon. And, in particular, we want to map new words to UNK and not add them to the lexicon.  So we need to change the default method for word2index.

In [195]:
word2index = defaultdict(lambda: UNK, word2index)

In [196]:
word2index["Gandalf"]

363

In [197]:
word2index["Ralphie"]

0

For a bigram model we need bigrams. So back to words.

We can get all the bigrams in the corpus by going back to the words and stepping through it a pair at a time.

In [198]:
for i in range(1, len(words)): print(words[i-1], words[i])

THE HOBBIT
HOBBIT OR
OR THERE
THERE AND
AND BACK
BACK AGAIN
AGAIN J.R.R.
J.R.R. TOLKIEN
TOLKIEN Chapter
Chapter I
I AN
AN UNEXPECTED
UNEXPECTED PARTY
PARTY In
In a
a hole
hole in
in the
the ground
ground there
there lived
lived a
a hobbit.
hobbit. Not
Not a
a nasty,
nasty, dirty,
dirty, wet
wet hole,
hole, filled
filled with
with the
the ends
ends of
of worms
worms and
and an
an oozy
oozy smell,
smell, nor
nor yet
yet a
a dry,
dry, bare,
bare, sandy
sandy hole
hole with
with nothing
nothing in
in it
it to
to sit
sit down
down on
on or
or to
to eat:
eat: it
it was
was a
a hobbit-hole,
hobbit-hole, and
and that
that means
means comfort.
comfort. It
It had
had a
a perfectly
perfectly round
round door
door like
like a
a porthole,
porthole, painted
painted green,
green, with
with a
a shiny
shiny yellow
yellow brass
brass knob
knob in
in the
the exact
exact middle.
middle. The
The door
door opened
opened on
on to
to a
a tube-shaped
tube-shaped hall
hall like
like a
a tunnel:
tunnel: a
a very

a quarter
quarter of
of what
what I
I have
have heard
heard about
about him,
him, and
and I
I have
have only
only heard
heard very
very little
little of
of all
all there
there is
is to
to hear,
hear, you
you would
would be
be prepared
prepared for
for any
any sort
sort of
of remarkable
remarkable tale.
tale. Tales
Tales and
and adventures
adventures sprouted
sprouted up
up all
all over
over the
the place
place wherever
wherever he
he went,
went, in
in the
the most
most extraordinary
extraordinary fashion.
fashion. He
He had
had not
not been
been down
down that
that way
way under
under The
The Hill
Hill for
for ages
ages and
and ages,
ages, not
not since
since his
his friend
friend the
the Old
Old Took
Took died,
died, in
in fact,
fact, and
and the
the hobbits
hobbits had
had almost
almost forgotten
forgotten what
what he
he looked
looked like.
like. He
He had
had been
been away
away over
over The
The Hill
Hill and
and across
across The
The Water
Water on
on businesses
businesses of
of 

up. He
He hung
hung his
his red
red one
one next
next to
to it,
it, and
and “Balin
“Balin at
at your
your service!”
service!” he
he said
said with
with his
his hand
hand on
on his
his breast.
breast. “Thank
“Thank you!”
you!” said
said Bilbo
Bilbo with
with a
a gasp.
gasp. It
It was
was not
not the
the correct
correct thing
thing to
to say,
say, but
but they
they have
have begun
begun to
to arrive
arrive had
had flustered
flustered him
him badly.
badly. He
He liked
liked visitors,
visitors, but
but he
he liked
liked to
to know
know them
them before
before they
they arrived,
arrived, and
and he
he preferred
preferred to
to ask
ask them
them himself.
himself. He
He had
had a
a horrible
horrible thought
thought that
that the
the cakes
cakes might
might run
run short,
short, and
and then
then he
he —
— as
as the
the host:
host: he
he knew
knew his
his duty
duty and
and stuck
stuck to
to it
it however
however painful
painful —
— he
he might
might have
have to
to go
go without.
without. “Com

the kitchen
kitchen trying
trying to
to see
see what
what they
they were
were doing.
doing. Then
Then they
they went
went back,
back, and
and found
found Thorin
Thorin with
with his
his feet
feet on
on the
the fender
fender smoking
smoking a
a pipe.
pipe. He
He was
was blowing
blowing the
the most
most enormous
enormous smoke-rings,
smoke-rings, and
and wherever
wherever he
he told
told one
one to
to go,
go, it
it went
went —
— up
up the
the chimney,
chimney, or
or behind
behind the
the clock
clock on
on the
the mantelpiece,
mantelpiece, or
or under
under the
the table,
table, or
or round
round and
and round
round the
the ceiling;
ceiling; but
but wherever
wherever it
it went
went it
it was
was not
not quick
quick enough
enough to
to escape
escape Gandalf.
Gandalf. Pop!
Pop! he
he sent
sent a
a smaller
smaller smoke-ring
smoke-ring from
from his
his short
short clay-pipe
clay-pipe straight
straight through
through each
each one
one of
of Thorin’s.
Thorin’s. Then
Then Gandalfs
Gandalfs 

went down
down a
a rabbit-hole,
rabbit-hole, and
and in
in this
this way
way the
the battle
battle was
was won
won and
and the
the game
game of
of Golf
Golf invented
invented at
at the
the same
same moment.
moment. In
In the
the meanwhile,
meanwhile, however,
however, Bullroarer’s
Bullroarer’s gentler
gentler descendant
descendant was
was reviving
reviving in
in the
the drawing-
drawing- room.
room. After
After a
a while
while and
and a
a drink
drink he
he crept
crept nervously
nervously to
to the
the door
door of
of the
the parlour.
parlour. This
This is
is what
what he
he heard,
heard, Gloin
Gloin speaking:
speaking: “Humph!”
“Humph!” (or
(or some
some snort
snort more
more or
or less
less like
like that).
that). “Will
“Will he
he do,
do, do
do you
you think?
think? It
It is
is all
all very
very well
well for
for Gandalf
Gandalf to
to talk
talk about
about this
this hobbit
hobbit being
being fierce,
fierce, but
but one
one shriek
shriek like
like that
that in
in a
a moment
moment of


like to
to know
know a
a bit
bit more
more about
about things,”
things,” said
said he,
he, feeling
feeling all
all confused
confused and
and a
a bit
bit shaky
shaky inside,
inside, but
but so
so far
far still
still Tookishly
Tookishly determined
determined to
to go
go on
on with
with things.
things. “I
“I mean
mean about
about the
the gold
gold and
and the
the dragon,
dragon, and
and all
all that,
that, and
and how
how it
it got
got there,
there, and
and who
who it
it belongs
belongs to,
to, and
and so
so on
on and
and further.”
further.” “Bless
“Bless me!”
me!” said
said Thorin,
Thorin, “haven’t
“haven’t you
you got
got a
a map?
map? and
and didn’t
didn’t you
you hear
hear our
our song?
song? and
and haven’t
haven’t we
we been
been talking
talking about
about all
all this
this for
for hours?”
hours?” “All
“All the
the same,
same, I
I should
should like
like it
it all
all plain
plain and
and clear,”
clear,” said
said he
he obstinately,
obstinately, putting
putting on
on his
his busines

the powers
powers of
of all
all the
the dwarves
dwarves put
put together,
together, if
if they
they could
could all
all be
be collected
collected again
again from
from the
the four
four corners
corners of
of the
the world.
world. The
The one
one thing
thing your
your father
father wished
wished was
was for
for his
his son
son to
to read
read the
the map
map and
and use
use the
the key.
key. The
The dragon
dragon and
and the
the Mountain
Mountain are
are more
more than
than big
big enough
enough tasks
tasks for
for you!”
you!” “Hear,
“Hear, hear!”
hear!” said
said Bilbo,
Bilbo, and
and accidentally
accidentally said
said it
it aloud.
aloud. “Hear
“Hear what?”
what?” they
they all
all said
said turning
turning suddenly
suddenly towards
towards him,
him, and
and he
he was
was so
so flustered
flustered that
that he
he answered
answered “Hear
“Hear what
what I
I have
have got
got to
to say!”
say!” “What’s
“What’s that?”
that?” they
they asked.
asked. “Well,
“Well, I
I should
should say
say 

far ahead
ahead were
were dreary
dreary hills,
hills, rising
rising higher
higher and
and higher,
higher, dark
dark with
with trees.
trees. On
On some
some of
of them
them were
were old
old castles
castles with
with an
an evil
evil look,
look, as
as if
if they
they had
had been
been built
built by
by wicked
wicked people.
people. Everything
Everything seemed
seemed gloomy,
gloomy, for
for the
the weather
weather that
that day
day had
had taken
taken a
a nasty
nasty turn.
turn. Mostly
Mostly it
it had
had been
been as
as good
good as
as May
May can
can be,
be, can
can be,
be, even
even in
in merry
merry tales,
tales, but
but now
now it
it was
was cold
cold and
and wet.
wet. In
In the
the Lone-
Lone- lands
lands they
they had
had been
been obliged
obliged to
to camp
camp when
when they
they could,
could, but
but at
at least
least it
it had
had been
been dry.
dry. “To
“To think
think it
it will
will soon
soon be
be June!”
June!” grumbled
grumbled Bilbo,
Bilbo, as
as he
he splashed
splashe

enough,” said
said a
a second.
second. “What
“What the
the 'ell
'ell William
William was
was a-thinkin'
a-thinkin' of
of to
to bring
bring us
us into
into these
these parts
parts at
at all,
all, beats
beats me
me —
— and
and the
the drink
drink runnin'
runnin' short,
short, what's
what's more,”
more,” he
he said
said jogging
jogging the
the elbow
elbow of
of William,
William, who
who was
was taking
taking a
a pull
pull at
at his
his jug.
jug. William
William choked.
choked. “Shut
“Shut yer
yer mouth!”
mouth!” he
he said
said as
as soon
soon as
as he
he could.
could. “Yer
“Yer can't
can't expect
expect folk
folk to
to stop
stop here
here for
for ever
ever just
just to
to be
be et
et by
by you
you and
and Bert.
Bert. You've
You've et
et a
a village
village and
and a
a half
half between
between yer,
yer, since
since we
we come
come down
down from
from the
the mountains.
mountains. How
How much
much more
more d'yer
d'yer want?
want? And
And time's
time's been
been up
up our
our way
way whe

neatly tied
tied up
up in
in sacks,
sacks, with
with three
three angry
angry trolls
trolls (and
(and two
two with
with burns
burns and
and bashes
bashes to
to remember)
remember) sitting
sitting by
by them,
them, arguing
arguing whether
whether they
they should
should roast
roast them
them slowly,
slowly, or
or mince
mince them
them fine
fine and
and boil
boil them,
them, or
or just
just sit
sit on
on them
them one
one by
by one
one and
and squash
squash them
them into
into jelly;
jelly; and
and Bilbo
Bilbo up
up in
in a
a bush,
bush, with
with his
his clothes
clothes and
and his
his skin
skin torn,
torn, not
not daring
daring to
to move
move for
for fear
fear they
they should
should hear
hear him.
him. It
It was
was just
just then
then that
that Gandalf
Gandalf came
came back.
back. But
But no
no one
one saw
saw him.
him. The
The trolls
trolls had
had just
just decided
decided to
to roast
roast the
the dwarves
dwarves now
now and
and eat
eat them
them later
later —
— that
that was
was

you know.
know. Please
Please be
be more
more careful,
careful, next
next time,
time, or
or we
we shall
shall never
never get
get anywhere!”
anywhere!” “Thank
“Thank you!”
you!” said
said Thorin.
Thorin. Chapter
Chapter III
III A
A SHORT
SHORT REST
REST They
They did
did not
not sing
sing or
or tell
tell stories
stories that
that day,
day, even
even though
though the
the weather
weather improved;
improved; nor
nor the
the next
next day,
day, nor
nor the
the day
day after.
after. They
They had
had begun
begun to
to feel
feel that
that danger
danger was
was not
not far
far away
away on
on either
either side.
side. They
They camped
camped under
under the
the stars,
stars, and
and their
their horses
horses had
had more
more to
to eat
eat than
than they
they had;
had; for
for there
there was
was plenty
plenty of
of grass,
grass, but
but there
there was
was not
not much
much in
in their
their bags,
bags, even
even with
with what
what they
they had
had got
got from
from the
the trolls.
trolls

words with
with these
these people
people that
that seemed
seemed to
to know
know his
his names
names and
and all
all about
about him,
him, although
although he
he had
had never
never seen
seen them
them before.
before. He
He thought
thought their
their opinion
opinion of
of his
his adventure
adventure might
might be
be interesting.
interesting. Elves
Elves know
know a
a lot
lot and
and are
are wondrous
wondrous folk
folk for
for news,
news, and
and know
know what
what is
is going
going on
on among
among the
the peoples
peoples of
of the
the land,
land, as
as quick
quick as
as water
water flows,
flows, or
or quicker.
quicker. But
But the
the dwarves
dwarves were
were all
all for
for supper
supper as
as soon
soon as
as possible
possible just
just then,
then, and
and would
would not
not stay.
stay. On
On they
they all
all went,
went, leading
leading their
their ponies,
ponies, till
till they
they were
were brought
brought to
to a
a good
good path
path and
and so
so at
at last
last to
to 

and deceptions
deceptions and
and led
led nowhere
nowhere or
or to
to bad
bad ends;
ends; and
and most
most of
of the
the passes
passes were
were infested
infested by
by evil
evil things
things and
and dreadful
dreadful dangers.
dangers. The
The dwarves
dwarves and
and the
the hobbit,
hobbit, helped
helped by
by the
the wise
wise advice
advice of
of Elrond
Elrond and
and the
the knowledge
knowledge and
and memory
memory of
of Gandalf,
Gandalf, took
took the
the right
right road
road to
to the
the right
right pass.
pass. Long
Long days
days after
after they
they had
had climbed
climbed out
out of
of the
the valley
valley and
and left
left the
the Last
Last Homely
Homely House
House miles
miles behind,
behind, they
they were
were still
still going
going up
up and
and up
up and
and up.
up. It
It was
was a
a hard
hard path
path and
and a
a dangerous
dangerous path,
path, a
a crooked
crooked way
way and
and a
a lonely
lonely and
and a
a long.
long. Now
Now they
they could
could look
look ba

and wider,
wider, and
and he
he was
was very
very afraid
afraid but
but could
could not
not call
call out
out or
or do
do anything
anything but
but lie
lie and
and look.
look. Then
Then he
he dreamed
dreamed that
that the
the floor
floor of
of the
the cave
cave was
was giving
giving way,
way, and
and he
he was
was slipping
slipping —
— beginning
beginning to
to fall
fall down,
down, down,
down, goodness
goodness knows
knows where
where to.
to. At
At that
that he
he woke
woke up
up with
with a
a horrible
horrible start,
start, and
and found
found that
that part
part of
of his
his dream
dream was
was true.
true. A
A crack
crack had
had opened
opened at
at the
the back
back of
of the
the cave,
cave, and
and was
was already
already a
a wide
wide passage.
passage. He
He was
was just
just in
in time
time to
to see
see the
the last
last of
of the
the ponies’
ponies’ tails
tails disappearing
disappearing into
into it.
it. Of
Of course
course he
he gave
gave a
a very
very loud
loud yell,
yell, 

“Slash them!
them! Beat
Beat them!
them! Bite
Bite them!
them! Gnash
Gnash them!
them! Take
Take them
them away
away to
to dark
dark holes
holes full
full of
of snakes,
snakes, and
and never
never let
let them
them see
see the
the light
light again!”
again!” He
He was
was in
in such
such a
a rage
rage that
that he
he jumped
jumped off
off his
his seat
seat and
and himself
himself rushed
rushed at
at Thorin
Thorin with
with his
his mouth
mouth open.
open. Just
Just at
at that
that moment
moment all
all the
the lights
lights in
in the
the cavern
cavern went
went out,
out, and
and the
the great
great fire
fire went
went off
off poof!
poof! into
into a
a tower
tower of
of blue
blue glowing
glowing smoke,
smoke, right
right up
up to
to the
the roof,
roof, that
that scattered
scattered piercing
piercing white
white sparks
sparks all
all among
among the
the goblins.
goblins. The
The yells
yells and
and yammering,
yammering, croaking,
croaking, jibbering
jibbering and
and jabbering;
jabbering;

all, and
and that
that shattered
shattered his
his hopes
hopes completely.
completely. Just
Just as
as well
well for
for him,
him, as
as he
he agreed
agreed when
when he
he came
came to
to his
his senses.
senses. Goodness
Goodness knows
knows what
what the
the striking
striking of
of matches
matches and
and the
the smell
smell of
of tobacco
tobacco would
would have
have brought
brought on
on him
him out
out of
of dark
dark holes
holes in
in that
that horrible
horrible place.
place. Still
Still at
at the
the moment
moment he
he felt
felt very
very crushed.
crushed. But
But in
in slapping
slapping all
all his
his pockets
pockets and
and feeling
feeling all
all round
round himself
himself for
for matches
matches his
his hand
hand came
came on
on the
the hilt
hilt of
of his
his little
little sword
sword —
— the
the little
little dagger
dagger that
that he
he got
got from
from the
the trolls,
trolls, and
and that
that he
he had
had quite
quite forgotten;
forgotten; nor
nor fortunately
fortu

taller than
than trees,
trees, Up,
Up, up
up it
it goes,
goes, And
And yet
yet never
never grows?
grows? “Easy!”
“Easy!” said
said Bilbo.
Bilbo. “Mountain,
“Mountain, I
I suppose.”
suppose.” “Does
“Does it
it guess
guess easy?
easy? It
It must
must have
have a
a competition
competition with
with us,
us, my
my preciouss!
preciouss! If
If precious
precious asks,
asks, and
and it
it doesn’t
doesn’t answer,
answer, we
we eats
eats it,
it, my
my preciousss.
preciousss. If
If it
it asks
asks us,
us, and
and we
we doesn’t
doesn’t answer,
answer, then
then we
we does
does what
what it
it wants,
wants, eh?
eh? We
We shows
shows it
it the
the way
way out,
out, yes!”
yes!” “All
“All right!”
right!” said
said Bilbo,
Bilbo, not
not daring
daring to
to disagree,
disagree, and
and nearly
nearly bursting
bursting his
his brain
brain to
to think
think of
of riddles
riddles that
that could
could save
save him
him from
from being
being eaten.
eaten. Thirty
Thirty white
white horses
horses on
on a
a red
r

pockets: fish-bones,
fish-bones, goblins’
goblins’ teeth,
teeth, wet
wet shells,
shells, a
a bit
bit of
of bat-wing,
bat-wing, a
a sharp
sharp stone
stone to
to sharpen
sharpen his
his fangs
fangs on,
on, and
and other
other nasty
nasty things.
things. He
He tried
tried to
to think
think what
what other
other people
people kept
kept in
in their
their pockets.
pockets. “Knife!”
“Knife!” he
he said
said at
at last.
last. “Wrong!”
“Wrong!” said
said Bilbo,
Bilbo, who
who had
had lost
lost his
his some
some time
time ago.
ago. “Last
“Last guess!”
guess!” Now
Now Gollum
Gollum was
was in
in a
a much
much worse
worse state
state than
than when
when Bilbo
Bilbo had
had asked
asked him
him the
the egg-
egg- question.
question. He
He hissed
hissed and
and spluttered
spluttered and
and rocked
rocked himself
himself backwards
backwards and
and forwards,
forwards, and
and slapped
slapped his
his feet
feet on
on the
the floor,
floor, and
and wriggled
wriggled and
and squirmed;
squirmed; but
but sti

he ran.
ran. What
What could
could it
it mean?
mean? Gollum
Gollum could
could see
see in
in the
the dark.
dark. Bilbo
Bilbo could
could see
see the
the light
light of
of his
his eyes
eyes palely
palely shining
shining even
even from
from behind.
behind. Painfully
Painfully he
he got
got up,
up, and
and sheathed
sheathed his
his sword,
sword, which
which was
was now
now glowing
glowing faintly
faintly again,
again, then
then very
very cautiously
cautiously he
he followed.
followed. There
There seemed
seemed nothing
nothing else
else to
to do.
do. It
It was
was no
no good
good crawling
crawling back
back down
down to
to Gollum’s
Gollum’s water.
water. Perhaps
Perhaps if
if he
he followed
followed him,
him, Gollum
Gollum might
might lead
lead him
him to
to some
some way
way of
of escape
escape without
without meaning
meaning to.
to. “Curse
“Curse it!
it! curse
curse it!
it! curse
curse it!”
it!” hissed
hissed Gollum.
Gollum. “Curse
“Curse the
the Baggins!
Baggins! It’s
It’s gone!
gone! W

suddenly he
he saw
saw the
the goblins:
goblins: goblins
goblins in
in full
full armour
armour with
with drawn
drawn swords
swords sitting
sitting just
just inside
inside the
the door,
door, and
and watching
watching it
it with
with wide
wide eyes,
eyes, and
and watching
watching the
the passage
passage that
that led
led to
to it.
it. They
They were
were aroused,
aroused, alert,
alert, ready
ready for
for anything.
anything. They
They saw
saw him
him sooner
sooner than
than he
he saw
saw them.
them. Yes,
Yes, they
they saw
saw him.
him. Whether
Whether it
it was
was an
an accident,
accident, or
or a
a last
last trick
trick of
of the
the ring
ring before
before it
it took
took a
a new
new master,
master, it
it was
was not
not on
on his
his finger.
finger. With
With yells
yells of
of delight
delight the
the goblins
goblins rushed
rushed upon
upon him.
him. A
A pang
pang of
of fear
fear and
and loss,
loss, like
like an
an echo
echo of
of Gollum’s
Gollum’s misery,
misery, smote
smote Bilbo

this. If
If they
they had
had still
still doubted
doubted that
that he
he was
was really
really a
a first-class
first-class burglar,
burglar, in
in spite
spite of
of Gandalf
Gandalf s
s words,
words, they
they doubted
doubted no
no longer.
longer. Balin
Balin was
was the
the most
most puzzled
puzzled of
of all;
all; but
but everyone
everyone said
said it
it was
was a
a very
very clever
clever bit
bit of
of work.
work. Indeed
Indeed Bilbo
Bilbo was
was so
so pleased
pleased with
with their
their praise
praise that
that he
he just
just chuckled
chuckled inside
inside and
and said
said nothing
nothing whatever
whatever about
about the
the ring;
ring; and
and when
when they
they asked
asked him
him how
how he
he did
did it,
it, he
he said:
said: “Oh,
“Oh, just
just crept
crept along,
along, you
you know
know —
— very
very carefully
carefully and
and quietly.”
quietly.” “Well,
“Well, It
It is
is the
the first
first time
time that
that even
even a
a mouse
mouse has
has crept
crept along
alon

it difficult
difficult to
to send
send stones
stones bouncing
bouncing down
down on
on our
our heads.”
heads.” The
The dwarves
dwarves (and
(and Bilbo)
Bilbo) were
were feeling
feeling far
far from
from happy,
happy, and
and were
were rubbing
rubbing their
their bruised
bruised and
and damaged
damaged legs
legs and
and feet.
feet. “Nonsense!
“Nonsense! We
We are
are going
going to
to turn
turn aside
aside here
here out
out of
of the
the path
path of
of the
the slide.
slide. We
We must
must be
be quick!
quick! Look
Look at
at the
the light!”
light!” The
The sun
sun had
had long
long gone
gone behind
behind the
the mountains.
mountains. Already
Already the
the shadows
shadows were
were deepening
deepening about
about them,
them, though
though far
far away
away through
through the
the trees
trees and
and over
over the
the black
black tops
tops of
of those
those growing
growing lower
lower down
down they
they could
could still
still see
see the
the evening
evening lights
lights on
on the
t

But now
now they
they had
had planned
planned with
with the
the goblins’
goblins’ help
help to
to come
come by
by night
night upon
upon some
some of
of the
the villages
villages nearest
nearest the
the mountains.
mountains. If
If their
their plan
plan had
had been
been carried
carried out,
out, there
there would
would have
have been
been none
none left
left there
there next
next day;
day; all
all would
would have
have been
been killed
killed except
except the
the few
few the
the goblins
goblins kept
kept from
from the
the wolves
wolves and
and carried
carried back
back as
as prisoners
prisoners to
to their
their caves.
caves. This
This was
was dreadful
dreadful talk
talk to
to listen
listen to,
to, not
not only
only because
because of
of the
the brave
brave woodmen
woodmen and
and their
their wives
wives and
and children,
children, but
but also
also because
because of
of the
the danger
danger which
which now
now threatened
threatened Gandalf
Gandalf and
and his
his friends.
friends. Th

goblins. That
That would
would have
have been
been the
the end
end of
of him,
him, though
though he
he would
would probably
probably have
have killed
killed many
many of
of them
them as
as he
he came
came hurtling
hurtling down
down like
like a
a thunderbolt.
thunderbolt. But
But he
he never
never leaped.
leaped. Just
Just at
at that
that moment
moment the
the Lord
Lord of
of the
the Eagles
Eagles swept
swept down
down from
from above,
above, seized
seized him
him in
in his
his talons,
talons, and
and was
was gone.
gone. There
There was
was a
a howl
howl of
of anger
anger and
and surprise
surprise from
from the
the goblins.
goblins. Loud
Loud cried
cried the
the Lord
Lord of
of the
the Eagles,
Eagles, to
to whom
whom Gandalf
Gandalf had
had now
now spoken.
spoken. Back
Back swept
swept the
the great
great birds
birds that
that were
were with
with him,
him, and
and down
down they
they came
came like
like huge
huge black
black shadows.
shadows. The
The wolves
wolves yammered
yammered and

back behind
behind them
them into
into the
the distance.
distance. He
He shut
shut his
his eyes
eyes again
again and
and held
held on
on tighter.
tighter. “Don’t
“Don’t pinch!”
pinch!” said
said his
his eagle.
eagle. “You
“You need
need not
not be
be frightened
frightened like
like a
a rabbit,
rabbit, even
even if
if you
you look
look rather
rather like
like one.
one. It
It is
is a
a fair
fair morning
morning with
with little
little wind.
wind. What
What is
is finer
finer than
than flying?”
flying?” Bilbo
Bilbo would
would have
have liked
liked to
to say:
say: “A
“A warm
warm bath
bath and
and late
late breakfast
breakfast on
on the
the lawn
lawn afterwards;”
afterwards;” but
but he
he thought
thought it
it better
better to
to say
say nothing
nothing at
at all,
all, and
and to
to let
let go
go his
his clutch
clutch just
just a
a tiny
tiny bit.
bit. After
After a
a good
good while
while the
the eagles
eagles must
must have
have seen
seen the
the point
point they
they were
were making
ma

enough yet
yet to
to have
have fallen
fallen to
to the
the ground.
ground. It
It was
was the
the middle
middle of
of the
the afternoon
afternoon before
before they
they noticed
noticed that
that great
great patches
patches of
of flowers
flowers had
had begun
begun to
to spring
spring up,
up, all
all the
the same
same kinds
kinds growing
growing together
together as
as if
if they
they had
had been
been planted.
planted. Especially
Especially there
there was
was clover,
clover, waving
waving patches
patches of
of cockscomb
cockscomb clover,
clover, and
and purple
purple clover,
clover, and
and wide
wide stretches
stretches of
of short
short white
white sweet
sweet honey-smelling
honey-smelling clover.
clover. There
There was
was a
a buzzing
buzzing and
and a
a whirring
whirring and
and a
a droning
droning in
in the
the air.
air. Bees
Bees were
were busy
busy everywhere.
everywhere. And
And such
such bees!
bees! Bilbo
Bilbo had
had never
never seen
seen anything
anything like
like them.
t

went on
on Gandalf,
Gandalf, “a
“a crack
crack at
at the
the back
back of
of the
the cave
cave opened;
opened; goblins
goblins came
came out
out and
and grabbed
grabbed the
the hobbit
hobbit and
and the
the dwarves
dwarves and
and our
our troop
troop of
of ponies
ponies “Troop
“Troop of
of ponies?
ponies? What
What were
were you
you —
— a
a travelling
travelling circus?
circus? Or
Or were
were you
you carrying
carrying lots
lots of
of goods?
goods? Or
Or do
do you
you always
always call
call six
six a
a troop?”
troop?” “O
“O no!
no! As
As a
a matter
matter of
of fact
fact there
there were
were more
more than
than six
six ponies,
ponies, for
for there
there were
were more
more than
than six
six of
of us
us —
— and
and well,
well, here
here are
are two
two more!”
more!” Just
Just at
at that
that moment
moment Balin
Balin and
and Dwalin
Dwalin appeared
appeared and
and bowed
bowed so
so low
low that
that their
their beards
beards swept
swept the
the stone
stone floor.
floor. The
The big
b

of gold
gold and
and silver
silver and
and jewels
jewels and
and the
the making
making of
of things
things by
by smith-craft,
smith-craft, and
and Beorn
Beorn did
did not
not appear
appear to
to care
care for
for such
such things:
things: there
there were
were no
no things
things of
of gold
gold or
or silver
silver in
in his
his hall,
hall, and
and few
few save
save the
the knives
knives were
were made
made of
of metal
metal at
at all.
all. They
They sat
sat long
long at
at the
the table
table with
with their
their wooden
wooden drinking-bowls
drinking-bowls filled
filled with
with mead.
mead. The
The dark
dark night
night came
came on
on outside.
outside. The
The fires
fires in
in the
the middle
middle of
of the
the hall
hall were
were built
built with
with fresh
fresh logs
logs and
and the
the torches
torches were
were put
put out,
out, and
and still
still they
they sat
sat in
in the
the light
light of
of the
the dancing
dancing flames
flames with
with the
the pillars
pillars of
of t

still!” he
he said.
said. He
He picked
picked up
up the
the hobbit
hobbit and
and laughed:
laughed: “Not
“Not eaten
eaten up
up by
by Wargs
Wargs or
or goblins
goblins or
or wicked
wicked bears
bears yet
yet I
I see”;
see”; and
and he
he poked
poked Mr.
Mr. Baggins’
Baggins’ waistcoat
waistcoat most
most disrespectfully.
disrespectfully. “Little
“Little bunny
bunny is
is getting
getting nice
nice and
and fat
fat again
again on
on bread
bread and
and honey,”
honey,” he
he chuckled.
chuckled. “Come
“Come and
and have
have some
some more!”
more!” So
So they
they all
all went
went to
to breakfast
breakfast with
with him.
him. Beorn
Beorn was
was most
most jolly
jolly for
for a
a change;
change; indeed
indeed he
he seemed
seemed to
to be
be in
in a
a splendidly
splendidly good
good humour
humour and
and set
set them
them all
all laughing
laughing with
with his
his funny
funny stories;
stories; nor
nor did
did they
they have
have to
to wonder
wonder long
long where
where he
he had
had been
b

howl of
of hunting
hunting wolves
wolves and
and the
the cries
cries of
of goblins.
goblins. Still
Still the
the next
next morning
morning dawned
dawned bright
bright and
and fair
fair again.
again. There
There was
was an
an autumn-like
autumn-like mist
mist white
white upon
upon the
the ground
ground and
and the
the air
air was
was chill,
chill, but
but soon
soon the
the sun
sun rose
rose red
red in
in the
the East
East and
and the
the mists
mists vanished,
vanished, and
and while
while the
the shadows
shadows were
were still
still long
long they
they were
were off
off again.
again. So
So they
they rode
rode now
now for
for two
two more
more days,
days, and
and all
all the
the while
while they
they saw
saw nothing
nothing save
save grass
grass and
and flowers
flowers and
and birds
birds and
and scattered
scattered trees,
trees, and
and occasionally
occasionally small
small herds
herds of
of red
red deer
deer browsing
browsing or
or sitting
sitting at
at noon
noon in
in the
the shade.


and wound
wound in
in and
and out
out among
among the
the trunks.
trunks. Soon
Soon the
the light
light at
at the
the gate
gate was
was like
like a
a little
little bright
bright hole
hole far
far behind,
behind, and
and the
the quiet
quiet was
was so
so deep
deep that
that their
their feet
feet seemed
seemed to
to thump
thump along
along while
while all
all the
the trees
trees leaned
leaned over
over them
them and
and listened.
listened. As
As their
their eyes
eyes became
became used
used to
to the
the dimness
dimness they
they could
could see
see a
a little
little way
way to
to either
either side
side in
in a
a sort
sort of
of darkened
darkened green
green glimmer.
glimmer. Occasionally
Occasionally a
a slender
slender beam
beam of
of sun
sun that
that had
had the
the luck
luck to
to slip
slip in
in through
through some
some opening
opening in
in the
the leaves
leaves far
far above,
above, and
and still
still more
more luck
luck in
in not
not being
being caught
caught in
in the
the ta

they saw
saw that
that it
it must
must have
have stuck
stuck in
in the
the branches.
branches. “Get
“Get in
in now,”
now,” said
said Fili,
Fili, “and
“and one
one of
of you
you haul
haul on
on the
the rope
rope that
that is
is stuck
stuck in
in a
a tree
tree on
on the
the other
other side.
side. One
One of
of the
the others
others must
must keep
keep hold
hold of
of the
the hook
hook we
we used
used at
at first,
first, and
and when
when we
we are
are safe
safe on
on the
the other
other side
side he
he can
can hook
hook it
it on,
on, and
and you
you can
can draw
draw the
the boat
boat back.”
back.” In
In this
this way
way they
they were
were all
all soon
soon on
on the
the far
far bank
bank safe
safe across
across the
the enchanted
enchanted stream.
stream. Dwalin
Dwalin had
had just
just scrambled
scrambled out
out with
with the
the coiled
coiled rope
rope on
on his
his arm,
arm, and
and Bombur
Bombur (still
(still grumbling)
grumbling) was
was getting
getting ready
ready to
to follow,

but these
these were
were not
not purple
purple at
at all,
all, they
they were
were a
a dark
dark dark
dark velvety
velvety black
black without
without any
any markings
markings to
to be
be seen.
seen. He
He looked
looked at
at the
the “black
“black emperors”
emperors” for
for a
a long
long time,
time, and
and enjoyed
enjoyed the
the feel
feel of
of the
the breeze
breeze in
in his
his hair
hair and
and on
on his
his face;
face; but
but at
at length
length the
the cries
cries of
of the
the dwarves,
dwarves, who
who were
were now
now simply
simply stamping
stamping with
with impatience
impatience down
down below,
below, reminded
reminded him
him of
of his
his real
real business.
business. It
It was
was no
no good.
good. Gaze
Gaze as
as much
much as
as he
he might,
might, he
he could
could see
see no
no end
end to
to the
the trees
trees and
and the
the leaves
leaves in
in any
any direction.
direction. His
His heart,
heart, that
that had
had been
been lightened
lightened by
by the
the sigh

not far
far away
away were
were scores
scores of
of twinkling
twinkling lights,
lights, and
and they
they heard
heard the
the voices
voices and
and the
the laughter
laughter quite
quite plainly.
plainly. They
They crept
crept slowly
slowly towards
towards them,
them, in
in a
a single
single line,
line, each
each touching
touching the
the back
back of
of the
the one
one in
in front.
front. When
When they
they got
got near
near Thorin
Thorin said:
said: “No
“No rushing
rushing forward
forward this
this time!
time! No
No one
one is
is to
to stir
stir from
from hiding
hiding till
till I
I say.
say. I
I shall
shall send
send Mr.
Mr. Baggins
Baggins alone
alone first
first to
to talk
talk to
to them.
them. They
They won’t
won’t be
be frightened
frightened of
of him
him —
— (‘What
(‘What about
about me
me of
of them?’
them?’ thought
thought Bilbo)
Bilbo) —
— and
and any
any way
way I
I hope
hope they
they won’t
won’t do
do anything
anything nasty
nasty to
to him.”
him.” When
When they
they go

then in
in the
the silence
silence and
and stillness
stillness of
of the
the wood
wood he
he realised
realised that
that these
these loathsome
loathsome creatures
creatures were
were speaking
speaking one
one to
to another.
another. Their
Their voices
voices were
were a
a sort
sort of
of thin
thin creaking
creaking and
and hissing,
hissing, but
but he
he could
could make
make out
out many
many of
of the
the words
words that
that they
they said.
said. They
They were
were talking
talking about
about the
the dwarves!
dwarves! “It
“It was
was a
a sharp
sharp struggle,
struggle, but
but worth
worth it,”
it,” said
said one.
one. “What
“What nasty
nasty thick
thick skins
skins they
they have
have to
to be
be sure,
sure, but
but I'll
I'll wager
wager there
there is
is good
good juice
juice inside.”
inside.” “Aye,
“Aye, they’ll
they’ll make
make fine
fine eating,
eating, when
when they’ve
they’ve hung
hung a
a bit,”
bit,” said
said another.
another. “Don’t
“Don’t hang
hang ’em
’em too
too long,

first bundle.
bundle. “Fili
“Fili or
or Kili,”
Kili,” he
he thought
thought by
by the
the tip
tip of
of a
a blue
blue hood
hood sticking
sticking out
out at
at the
the top.
top. “Most
“Most likely
likely Fili,"
Fili," he
he thought
thought by
by the
the tip
tip of
of a
a long
long nose
nose poking
poking out
out of
of the
the winding
winding threads.
threads. He
He managed
managed by
by leaning
leaning over
over to
to cut
cut most
most of
of the
the strong
strong sticky
sticky threads
threads that
that bound
bound him
him round,
round, and
and then,
then, sure
sure enough,
enough, with
with a
a kick
kick and
and a
a struggle
struggle most
most of
of Fili
Fili emerged.
emerged. I
I am
am afraid
afraid Bilbo
Bilbo actually
actually laughed
laughed at
at the
the sight
sight of
of him
him jerking
jerking his
his stiff
stiff arms
arms and
and legs
legs as
as he
he danced
danced on
on the
the spider-string
spider-string under
under his
his armpits,
armpits, just
just like
like one
one of
of 

with the
the ring
ring in
in its
its proper
proper place.
place. But
But after
after a
a time
time the
the light
light began
began to
to fail,
fail, and
and then
then other
other questions
questions were
were asked.
asked. Where
Where were
were they,
they, and
and where
where was
was their
their path,
path, and
and where
where was
was there
there any
any food,
food, and
and what
what were
were they
they going
going to
to do
do next?
next? These
These questions
questions they
they asked
asked over
over and
and over
over again,
again, and
and it
it was
was from
from little
little Bilbo
Bilbo that
that they
they seemed
seemed to
to expect
expect to
to get
get the
the answers.
answers. From
From which
which you
you can
can see
see that
that they
they had
had changed
changed their
their opinion
opinion of
of Mr.
Mr. Baggins
Baggins very
very much,
much, and
and had
had begun
begun to
to have
have a
a great
great respect
respect for
for him
him (as
(as Gandalf
Gandalf had
had said
said they


OUT OF
OF BOND
BOND The
The day
day after
after the
the battle
battle with
with the
the spiders
spiders Bilbo
Bilbo and
and the
the dwarves
dwarves made
made one
one last
last despairing
despairing effort
effort to
to find
find a
a way
way out
out before
before they
they died
died of
of hunger
hunger and
and thirst.
thirst. They
They got
got up
up and
and staggered
staggered on
on in
in the
the direction
direction which
which eight
eight out
out of
of the
the thirteen
thirteen of
of them
them guessed
guessed to
to be
be the
the one
one in
in which
which the
the path
path lay;
lay; but
but they
they never
never found
found out
out if
if they
they were
were right.
right. Such
Such day
day as
as there
there ever
ever was
was in
in the
the forest
forest was
was fading
fading once
once more
more into
into the
the blackness
blackness of
of night,
night, when
when suddenly
suddenly out
out sprang
sprang the
the light
light of
of many
many torches
torches all
all round
round them,
them, like
l

the chief
chief of
of the
the dwarves.
dwarves. Thorin
Thorin was
was too
too wretched
wretched to
to be
be angry
angry any
any longer
longer at
at his
his misfortunes,
misfortunes, and
and was
was even
even beginning
beginning to
to think
think of
of telling
telling the
the king
king all
all about
about his
his treasure
treasure and
and his
his quest
quest (which
(which shows
shows how
how low-
low- spirited
spirited he
he had
had become),
become), when
when he
he heard
heard Bilbo’s
Bilbo’s little
little voice
voice at
at his
his keyhole.
keyhole. He
He could
could hardly
hardly believe
believe his
his ears.
ears. Soon
Soon however
however he
he made
made up
up his
his mind
mind that
that he
he could
could not
not be
be mistaken,
mistaken, and
and he
he came
came to
to the
the door
door and
and had
had a
a long
long whispered
whispered talk
talk with
with the
the hobbit
hobbit on
on the
the other
other side.
side. So
So it
it was
was that
that Bilbo
Bilbo was
was able
able to
to take

bumped into
into another,
another, or
or grunted
grunted or
or whispered
whispered in
in the
the dark.
dark. “Drat
“Drat this
this dwarvish
dwarvish racket!”
racket!” he
he said
said to
to himself.
himself. But
But all
all went
went well,
well, and
and they
they met
met no
no guards.
guards. As
As a
a matter
matter of
of fact
fact there
there was
was a
a great
great autumn
autumn feast
feast in
in the
the woods
woods that
that night,
night, and
and in
in the
the halls
halls above.
above. Nearly
Nearly all
all the
the king’s
king’s folk
folk were
were merrymaking.
merrymaking. At
At last
last after
after much
much blundering
blundering they
they came
came to
to Thorin’s
Thorin’s dungeon,
dungeon, far
far down
down in
in a
a deep
deep place
place and
and fortunately
fortunately not
not far
far from
from the
the cellars.
cellars. “Upon
“Upon my
my word!”
word!” said
said Thorin,
Thorin, when
when Bilbo
Bilbo whispered
whispered to
to him
him to
to come
come out
out and
and join
join his
h

once they
they were
were out
out of
of the
the wood.
wood. While
While all
all these
these thoughts
thoughts were
were passing
passing through
through his
his mind,
mind, the
the elves
elves being
being very
very merry
merry began
began to
to sing
sing a
a song
song round
round the
the river-door.
river-door. Some
Some had
had already
already gone
gone to
to haul
haul on
on the
the ropes
ropes which
which pulled
pulled up
up the
the portcullis
portcullis at
at the
the water-gate
water-gate so
so as
as to
to let
let out
out the
the barrels
barrels as
as soon
soon as
as they
they were
were all
all afloat
afloat below.
below. Down
Down the
the swift
swift dark
dark stream
stream you
you go
go Back
Back to
to lands
lands you
you once
once did
did know/.
know/. Leave
Leave the
the halls
halls and
and caverns
caverns deep,
deep, Leave
Leave the
the northern
northern mountains
mountains steep,
steep, Where
Where the
the forest
forest wide
wide and
and dim
dim Stoops
Stoops in
in shadow
shadow

some grumbled.
grumbled. “They
“They float
float too
too deep
deep —
— some
some of
of these
these are
are never
never empty.
empty. If
If they
they had
had come
come ashore
ashore in
in the
the daylight,
daylight, we
we might
might have
have had
had a
a look
look inside,”
inside,” they
they said.
said. “No
“No time
time now!”
now!” cried
cried the
the raftman.
raftman. “Shove
“Shove off!”
off!” And
And off
off they
they went
went at
at last,
last, slowly
slowly at
at first,
first, until
until they
they had
had passed
passed the
the point
point of
of rock
rock where
where other
other elves
elves stood
stood to
to fend
fend them
them off
off with
with poles,
poles, and
and then
then quicker
quicker and
and quicker
quicker as
as they
they caught
caught the
the main
main stream
stream and
and went
went sailing
sailing away
away down,
down, down
down towards
towards the
the Lake.
Lake. They
They had
had escaped
escaped the
the dungeons
dungeons of
of the
the king
king and
and were
were thr

shoulder of
of rock
rock into
into the
the little
little bay
bay of
of Lake-town.
Lake-town. There
There it
it was
was moored
moored not
not far
far from
from the
the shoreward
shoreward head
head of
of the
the great
great bridge.
bridge. Soon
Soon men
men would
would come
come up
up from
from the
the South
South and
and take
take some
some of
of the
the casks
casks away,
away, and
and others
others they
they would
would fill
fill with
with goods
goods they
they had
had brought
brought to
to be
be taken
taken back
back up
up the
the stream
stream to
to the
the Wood-elves’
Wood-elves’ home.
home. In
In the
the meanwhile
meanwhile the
the barrels
barrels were
were left
left afloat
afloat while
while the
the elves
elves of
of the
the raft
raft and
and the
the boatmen
boatmen went
went to
to feast
feast in
in Lake-town.
Lake-town. They
They would
would have
have been
been surprised,
surprised, if
if they
they could
could have
have seen
seen what
what happened
happened down
down by
by the
t

wished for
for no
no enmity
enmity with
with him,
him, nor
nor did
did he
he think
think much
much of
of old
old songs,
songs, giving
giving his
his mind
mind to
to trade
trade and
and tolls,
tolls, to
to cargoes
cargoes and
and gold,
gold, to
to which
which habit
habit he
he owed
owed his
his position.
position. Others
Others were
were of
of different
different mind,
mind, however,
however, and
and quickly
quickly the
the matter
matter was
was settled
settled without
without him.
him. The
The news
news had
had spread
spread from
from the
the doors
doors of
of the
the hall
hall like
like fire
fire through
through all
all the
the town.
town. People
People were
were shouting
shouting inside
inside the
the hall
hall and
and outside
outside it.
it. The
The quays
quays were
were thronged
thronged with
with hurrying
hurrying feet.
feet. Some
Some began
began to
to sing
sing snatches
snatches of
of old
old songs
songs concerning
concerning the
the return
return of
of the
the King
King under
u

River Running,
Running, and
and drawing
drawing ever
ever nearer
nearer and
and nearer
nearer to
to a
a great
great spur
spur of
of the
the Mountain
Mountain that
that was
was flung
flung out
out southwards
southwards towards
towards them.
them. It
It was
was a
a weary
weary journey,
journey, and
and a
a quiet
quiet and
and stealthy
stealthy one.
one. There
There was
was no
no laughter
laughter or
or song
song or
or sound
sound of
of harps,
harps, and
and the
the pride
pride and
and hopes
hopes which
which had
had stirred
stirred in
in their
their hearts
hearts at
at the
the singing
singing of
of old
old songs
songs by
by the
the lake
lake died
died away
away to
to a
a plodding
plodding gloom.
gloom. They
They knew
knew that
that they
they were
were drawing
drawing near
near to
to the
the end
end of
of their
their journey,
journey, and
and that
that it
it might
might be
be a
a very
very horrible
horrible end.
end. The
The land
land about
about them
them grew
grew bleak
bleak and
and ba

and tread
tread on
on my
my beard,
beard, and
and then
then you
you would
would be
be thirteen
thirteen again.
again. And
And the
the knotted
knotted ropes
ropes are
are too
too slender
slender for
for my
my weight.”
weight.” Luckily
Luckily for
for him
him that
that was
was not
not true,
true, as
as you
you will
will see.
see. In
In the
the meanwhile
meanwhile some
some of
of them
them explored
explored the
the ledge
ledge beyond
beyond the
the opening
opening and
and found
found a
a path
path that
that led
led higher
higher and
and higher
higher on
on to
to the
the mountain;
mountain; but
but they
they did
did not
not dare
dare to
to venture
venture very
very far
far that
that way,
way, nor
nor was
was there
there much
much use
use in
in it.
it. Out
Out up
up there
there a
a silence
silence reigned,
reigned, broken
broken by
by no
no bird
bird or
or sound
sound except
except that
that of
of the
the wind
wind in
in the
the crannies
crannies of
of stone.
stone. They
They spoke
spoke lo

it seemed
seemed centuries
centuries ago
ago —
— “but
“but anyway
anyway I
I think
think I
I will
will go
go and
and have
have a
a peep
peep at
at once
once and
and get
get it
it over.
over. Now
Now who
who is
is coming
coming with
with me?”
me?” He
He did
did not
not expect
expect a
a chorus
chorus of
of volunteers,
volunteers, so
so he
he was
was not
not disappointed.
disappointed. Fili
Fili and
and Kili
Kili looked
looked uncomfortable
uncomfortable and
and stood
stood on
on one
one leg,
leg, but
but the
the others
others made
made no
no pretence
pretence of
of offering
offering —
— except
except old
old Balin,
Balin, the
the lookout
lookout man,
man, who
who was
was rather
rather fond
fond of
of the
the hobbit.
hobbit. He
He said
said he
he would
would come
come inside
inside at
at least
least and
and perhaps
perhaps a
a bit
bit of
of the
the way
way too,
too, ready
ready to
to call
call for
for help
help if
if necessary.
necessary. The
The most
most that
that can
can be
be said
sa

they know
know it
it to
to an
an ounce
ounce as
as a
a rule,
rule, especially
especially after
after long
long possession;
possession; and
and Smaug
Smaug was
was no
no exception.
exception. He
He had
had passed
passed from
from an
an uneasy
uneasy dream
dream (in
(in which
which a
a warrior,
warrior, altogether
altogether insignificant
insignificant in
in size
size but
but provided
provided with
with a
a bitter
bitter sword
sword and
and great
great courage,
courage, figured
figured most
most unpleasantly)
unpleasantly) to
to a
a doze,
doze, and
and from
from a
a doze
doze to
to wide
wide waking.
waking. There
There was
was a
a breath
breath of
of strange
strange air
air in
in his
his cave.
cave. Could
Could there
there be
be a
a draught
draught from
from that
that little
little hole?
hole? He
He had
had never
never felt
felt quite
quite happy
happy about
about it,
it, though
though it
it was
was so
so small,
small, and
and now
now he
he glared
glared at
at it
it in
in suspicion
suspi

he started,
started, but
but it
it was
was as
as dark
dark as
as night
night in
in the
the tunnel.
tunnel. The
The light
light from
from the
the door,
door, almost
almost closed,
closed, soon
soon faded
faded as
as he
he went
went down.
down. So
So silent
silent was
was his
his going
going that
that smoke
smoke on
on a
a gentle
gentle wind
wind could
could hardly
hardly have
have surpassed
surpassed it,
it, and
and he
he was
was inclined
inclined to
to feel
feel a
a bit
bit proud
proud of
of himself
himself as
as he
he drew
drew near
near the
the lower
lower door.
door. There
There was
was only
only the
the veryfaintestglowto
veryfaintestglowto be
be seen.
seen. “Old
“Old Smaug
Smaug is
is weary
weary and
and asleep,”
asleep,” he
he thought.
thought. “He
“He can’t
can’t see
see me
me and
and he
he won’t
won’t hear
hear me.
me. Cheer
Cheer up
up Bilbo!”
Bilbo!” He
He had
had forgotten
forgotten or
or had
had never
never heard
heard about
about dragons’
dragons’ sense
sense of
of smell.


floor to
to ceiling
ceiling like
like scarlet
scarlet lightning.
lightning. ‘‘Revenge!
‘‘Revenge! The
The King
King under
under the
the Mountain
Mountain is
is dead
dead and
and where
where are
are his
his kin
kin that
that dare
dare seek
seek revenge?
revenge? Girion
Girion Lord
Lord of
of Dale
Dale is
is dead,
dead, and
and I
I have
have eaten
eaten his
his people
people like
like a
a wolf
wolf among
among sheep,
sheep, and
and where
where are
are his
his sons’
sons’ sons
sons that
that dare
dare approach
approach me?
me? I
I kill
kill where
where I
I wish
wish and
and none
none dare
dare resist.
resist. I
I laid
laid low
low the
the warriors
warriors of
of old
old and
and their
their like
like is
is not
not in
in the
the world
world today.
today. Then
Then I
I was
was but
but young
young and
and tender.
tender. Now
Now I
I am
am old
old and
and strong,
strong, strong,
strong, strong,
strong, Thief
Thief in
in the
the Shadows!”
Shadows!” he
he gloated.
gloated. “My
“My armour
armour 

that they
they had
had never
never thought
thought at
at all
all about
about what
what would
would happen
happen after
after the
the treasure
treasure had
had been
been won.
won. “We
“We knew
knew it
it would
would be
be a
a desperate
desperate venture,”
venture,” said
said Thorin,
Thorin, “and
“and we
we know
know that
that still;
still; and
and I
I still
still think
think that
that when
when we
we have
have won
won it
it will
will be
be time
time enough
enough to
to think
think what
what to
to do
do about
about it.
it. As
As for
for your
your share,
share, Mr.
Mr. Baggins,
Baggins, I
I assure
assure you
you we
we are
are more
more than
than grateful
grateful and
and you
you shall
shall choose
choose your
your own
own fourteenth,
fourteenth, as
as soon
soon as
as we
we have
have anything
anything to
to divide.
divide. I
I am
am sorry
sorry if
if you
you are
are worried
worried about
about transport,
transport, and
and I
I admit
admit the
the difficulties
difficulties are
are great
gre

Oin and
and Gloin
Gloin have
have not
not lost
lost their
their tinder-boxes,
tinder-boxes, perhaps
perhaps we
we can
can make
make a
a little
little light,
light, and
and have
have a
a look
look round
round before
before the
the luck
luck turns.”
turns.” “Light!”
“Light!” he
he cried.
cried. “Can
“Can anybody
anybody make
make a
a light?”
light?” The
The dwarves,
dwarves, of
of course,
course, were
were very
very alarmed
alarmed when
when Bilbo
Bilbo fell
fell forward
forward down
down the
the step
step with
with a
a bump
bump into
into the
the hall,
hall, and
and they
they sat
sat huddled
huddled just
just where
where he
he had
had left
left them
them at
at the
the end
end of
of the
the tunnel.
tunnel. “Sh!
“Sh! sh!”
sh!” they
they hissed,
hissed, when
when they
they heard
heard his
his voice;
voice; and
and though
though that
that helped
helped the
the hobbit
hobbit to
to find
find out
out where
where they
they were,
were, it
it was
was some
some time
time before
before he
he could


sat down
down on
on the
the floor;
floor; and
and he
he began
began to
to wonder
wonder nervously
nervously what
what the
the end
end of
of it
it all
all would
would be.
be. “I
“I would
would give
give a
a good
good many
many of
of these
these precious
precious goblets,”
goblets,” he
he thought,
thought, “for
“for a
a drink
drink of
of something
something cheering
cheering out
out of
of one
one of
of Beorn’s
Beorn’s wooden
wooden bowls!”
bowls!” “Thorin!”
“Thorin!” he
he cried
cried aloud.
aloud. “What
“What next?
next? We
We are
are armed,
armed, but
but what
what good
good has
has any
any armour
armour ever
ever been
been before
before against
against Smaug
Smaug the
the Dreadful?
Dreadful? This
This treasure
treasure is
is not
not yet
yet won
won back.
back. We
We are
are not
not looking
looking for
for gold
gold yet,
yet, but
but for
for a
a way
way of
of escape;
escape; and
and we
we have
have tempted
tempted luck
luck too
too long!”
long!” “You
“You speak
speak the
the truth!”
tr

the Mountain.
Mountain. But
But there
there seemed
seemed small
small need
need for
for watching
watching in
in the
the days
days of
of our
our prosperity,
prosperity, and
and the
the guards
guards were
were made
made over
over comfortable,
comfortable, perhaps
perhaps —
— otherwise
otherwise we
we might
might have
have had
had longer
longer warning
warning of
of the
the coming
coming of
of the
the dragon,
dragon, and
and things
things might
might have
have been
been different.
different. Still,
Still, here
here we
we can
can now
now lie
lie hid
hid and
and sheltered
sheltered for
for a
a while,
while, and
and can
can see
see much
much without
without being
being seen.”
seen.” "Not
"Not much
much use,
use, if
if we
we have
have been
been seen
seen coming
coming here,”
here,” said
said Dori,
Dori, who
who was
was always
always looking
looking up
up towards
towards the
the Mountain’s
Mountain’s peak,
peak, as
as if
if he
he expected
expected to
to see
see Smaug
Smaug perched
perched ther

was of
of the
the race
race of
of Dale.
Dale. “Wait!
“Wait! Wait!”
Wait!” it
it said
said to
to him.
him. “The
“The moon
moon is
is rising.
rising. Look
Look for
for the
the hollow
hollow of
of the
the left
left breast
breast as
as he
he flies
flies and
and turns
turns above
above you!”
you!” And
And while
while Bard
Bard paused
paused in
in wonder
wonder it
it told
told him
him of
of tidings
tidings up
up in
in the
the Mountain
Mountain and
and of
of all
all that
that it
it had
had heard.
heard. Then
Then Bard
Bard drew
drew his
his bow-string
bow-string to
to his
his ear.
ear. The
The dragon
dragon was
was circling
circling back,
back, flying
flying low,
low, and
and as
as he
he came
came the
the moon
moon rose
rose above
above the
the eastern
eastern shore
shore and
and silvered
silvered his
his great
great wings.
wings. “Arrow!”
“Arrow!” said
said the
the bowman.
bowman. “Black
“Black arrow!
arrow! I
I have
have saved
saved you
you to
to the
the last.
last. You
You have
have never


The Elvenking
Elvenking had
had received
received news
news from
from his
his own
own messengers
messengers and
and from
from the
the birds
birds that
that loved
loved his
his folk,
folk, and
and already
already knew
knew much
much of
of what
what had
had happened.
happened. Very
Very great
great indeed
indeed was
was the
the commotion
commotion among
among all
all things
things with
with wings
wings that
that dwelt
dwelt on
on the
the borders
borders of
of the
the Desolation
Desolation of
of the
the Dragon.
Dragon. The
The air
air was
was filled
filled with
with circling
circling flocks,
flocks, and
and their
their swift-flying
swift-flying messengers
messengers flew
flew here
here and
and there
there across
across the
the sky.
sky. Above
Above the
the borders
borders of
of the
the Forest
Forest there
there was
was whistling,
whistling, crying
crying and
and piping.
piping. Far
Far over
over Mirkwood
Mirkwood tidings
tidings spread:
spread: “Smaug
“Smaug is
is dead!”
dead!” Leaves
Lea

for they
they are
are homeless
homeless and
and many
many have
have died,
died, and
and Smaug
Smaug has
has destroyed
destroyed their
their town.
town. They
They too
too think
think to
to find
find amends
amends from
from your
your treasure,
treasure, whether
whether you
you are
are alive
alive or
or dead.
dead. “Your
“Your own
own wisdom
wisdom must
must decide
decide your
your course;
course; but
but thirteen
thirteen is
is small
small remnant
remnant of
of the
the great
great folk
folk of
of Durin
Durin that
that once
once dwelt
dwelt here,
here, and
and now
now are
are scattered
scattered far.
far. If
If you
you will
will listen
listen to
to my
my counsel,
counsel, you
you will
will not
not trust
trust the
the Master
Master of
of the
the Lake-men,
Lake-men, but
but rather
rather him
him that
that shot
shot the
the dragon
dragon with
with his
his bow.
bow. Bard
Bard is
is he,
he, of
of the
the race
race of
of Dale,
Dale, of
of the
the line
line of
of Girion;
Girion; he
he is
is a
a 

the gates
gates of
of Thorin
Thorin son
son of
of Thrain,
Thrain, King
King under
under the
the Mountain?”
Mountain?” This
This time
time he
he was
was answered.
answered. A
A tall
tall man
man stood
stood forward,
forward, dark
dark of
of hair
hair and
and grim
grim of
of face,
face, and
and he
he cried:
cried: “Hail
“Hail Thorin!
Thorin! Why
Why do
do you
you fence
fence yourself
yourself like
like a
a robber
robber in
in his
his hold?
hold? We
We are
are not
not yet
yet foes,
foes, and
and we
we rejoice
rejoice that
that you
you are
are alive
alive beyond
beyond our
our hope.
hope. We
We came
came expecting
expecting to
to find
find none
none living
living here;
here; yet
yet now
now that
that we
we are
are met
met there
there is
is matter
matter for
for a
a parley
parley and
and a
a council.”
council.” “Who
“Who are
are you,
you, and
and of
of what
what would
would you
you parley?”
parley?” “I
“I am
am Bard,
Bard, and
and by
by my
my hand
hand was
was the
the dragon
dragon slain
sl

for the
the feel
feel of
of a
a strong
strong drink
drink in
in my
my throat,
throat, and
and for
for a
a soft
soft bed
bed after
after a
a good
good supper!”
supper!” “I
“I can’t
can’t give
give you
you those,
those, while
while the
the siege
siege is
is going
going on.
on. But
But it
it is
is long
long since
since I
I watched,
watched, and
and I
I will
will take
take your
your turn
turn for
for you,
you, if
if you
you like.
like. There
There is
is no
no sleep
sleep in
in me
me tonight.”
tonight.” “You
“You are
are a
a good
good fellow,
fellow, Mr.
Mr. Baggins,
Baggins, and
and I
I will
will take
take your
your offer
offer kindly.
kindly. If
If there
there should
should be
be anything
anything to
to note,
note, rouse
rouse me
me first,
first, mind
mind you!
you! I
I will
will lie
lie in
in the
the inner
inner chamber
chamber to
to the
the left,
left, not
not far
far away.”
away.” “Off
“Off you
you go!”
go!” said
said Bilbo.
Bilbo. “I
“I will
will wake
wake you
you at
at midnight,
midn

he said
said farewell
farewell to
to the
the elves
elves and
and climbed
climbed carefully
carefully back
back towards
towards the
the Gate.
Gate. Great
Great weariness
weariness began
began to
to come
come over
over him;
him; but
but it
it was
was well
well before
before midnight
midnight when
when he
he clambered
clambered up
up the
the rope
rope again
again —
— it
it was
was still
still where
where he
he had
had left
left it.
it. He
He untied
untied it
it and
and hid
hid it,
it, and
and then
then he
he sat
sat down
down on
on the
the wall
wall and
and wondered
wondered anxiously
anxiously what
what would
would happen
happen next.
next. At
At midnight
midnight he
he woke
woke up
up Bombur;
Bombur; and
and then
then in
in turn
turn rolled
rolled himself
himself up
up in
in his
his corner,
corner, without
without listening
listening to
to the
the old
old dwarfs
dwarfs thanks
thanks (which
(which he
he felt
felt he
he had
had hardly
hardly earned).
earned). He
He was
was soon
soon fast


make way
way or
or we
we shall
shall fight
fight you!”
you!” They
They meant
meant to
to push
push on
on between
between the
the Mountain
Mountain and
and the
the loop
loop of
of the
the river;
river; for
for the
the narrow
narrow land
land there
there did
did not
not seem
seem to
to be
be strongly
strongly guarded.
guarded. Bard,
Bard, of
of course,
course, refused
refused to
to allow
allow the
the dwarves
dwarves to
to go
go straight
straight on
on to
to the
the Mountain.
Mountain. He
He was
was determined
determined to
to wait
wait until
until the
the gold
gold and
and silver
silver had
had been
been brought
brought out
out in
in exchange
exchange for
for the
the Arkenstone;
Arkenstone; for
for he
he did
did not
not believe
believe that
that this
this would
would be
be done,
done, if
if once
once the
the fortress
fortress was
was manned
manned with
with so
so large
large and
and warlike
warlike a
a company.
company. They
They had
had brought
brought with
with them
them a
a great
gre

“Moria!” and
and “Dain,
“Dain, Dain!”
Dain!” the
the dwarves
dwarves of
of the
the Iron
Iron Hills
Hills plunged
plunged in,
in, wielding
wielding their
their mattocks,
mattocks, upon
upon the
the other
other side;
side; and
and beside
beside them
them came
came the
the men
men of
of the
the Lake
Lake with
with long
long swords.
swords. Panic
Panic came
came upon
upon the
the Goblins;
Goblins; and
and even
even as
as they
they turned
turned to
to meet
meet this
this new
new attack,
attack, the
the elves
elves charged
charged again
again with
with renewed
renewed numbers.
numbers. Already
Already many
many of
of the
the goblins
goblins were
were flying
flying back
back down
down the
the river
river to
to escape
escape from
from the
the trap;
trap; and
and many
many of
of their
their own
own wolves
wolves were
were turning
turning upon
upon them
them and
and rending
rending the
the dead
dead and
and the
the wounded.
wounded. Victory
Victory seemed
seemed at
at hand,
hand, when
when a
a c

the host.
host. When
When Gandalf
Gandalf saw
saw Bilbo,
Bilbo, he
he was
was delighted.
delighted. “Baggins!”
“Baggins!” he
he exclaimed.
exclaimed. “Well
“Well I
I never!
never! Alive
Alive after
after all
all —
— I
I am
am glad!
glad! I
I began
began to
to wonder
wonder if
if even
even your
your luck
luck would
would see
see you
you through!
through! A
A terrible
terrible business,
business, and
and it
it nearly
nearly was
was disastrous.
disastrous. But
But other
other news
news can
can wait.
wait. Come!”
Come!” he
he said
said more
more gravely.
gravely. “You
“You are
are called
called for;”
for;” and
and leading
leading the
the hobbit
hobbit he
he took
took him
him within
within the
the tent.
tent. “Hail!
“Hail! Thorin,”
Thorin,” he
he said
said as
as he
he entered.
entered. “I
“I have
have brought
brought him.”
him.” There
There indeed
indeed lay
lay Thorin
Thorin Oakenshield,
Oakenshield, wounded
wounded with
with many
many wounds,
wounds, and
and his
his rent
rent armour
armou

Mountain he
he added:
added: “Farewell
“Farewell Thorin
Thorin Oakenshield!
Oakenshield! And
And Fili
Fili and
and Kili!
Kili! May
May your
your memory
memory never
never fade!”
fade!” Then
Then the
the dwarves
dwarves bowed
bowed low
low before
before their
their Gate,
Gate, but
but words
words stuck
stuck in
in their
their throats.
throats. “Good-
“Good- bye
bye and
and good
good luck,
luck, wherever
wherever you
you fare!”
fare!” said
said Balin
Balin at
at last.
last. “If
“If ever
ever you
you visit
visit us
us again,
again, when
when our
our halls
halls are
are made
made fair
fair once
once more,
more, then
then the
the feast
feast shall
shall indeed
indeed be
be splendid!”
splendid!” “If
“If ever
ever you
you are
are passing
passing my
my way,”
way,” said
said Bilbo,
Bilbo, “don’t
“don’t wait
wait to
to knock!
knock! Tea
Tea is
is at
at four;
four; but
but any
any of
of you
you are
are welcome
welcome at
at anytime!”
anytime!” Then
Then he
he turned
turned away.
away. The
The elf

Lullaby! Alder
Alder and
and Willow/
Willow/ Sigh
Sigh no
no more
more Pine,
Pine, till
till the
the wind
wind of
of the
the mom!
mom! Fall
Fall Moon!
Moon! Dark
Dark be
be the
the land!
land! Hush!
Hush! Hush!
Hush! Oak,
Oak, Ash,
Ash, and
and Thom!
Thom! Hushed
Hushed be
be all
all water,
water, till
till dawn
dawn is
is at
at hand!
hand! “Well,
“Well, Merry
Merry People!”
People!” said
said Bilbo
Bilbo looking
looking out.
out. “What
“What time
time by
by the
the moon
moon is
is this?
this? Your
Your lullaby
lullaby would
would waken
waken a
a drunken
drunken goblin!
goblin! Yet
Yet I
I thank
thank you.”
you.” “And
“And your
your snores
snores would
would waken
waken a
a stone
stone dragon
dragon —
— yet
yet we
we thank
thank you,”
you,” they
they answered
answered with
with laughter.
laughter. “It
“It is
is drawing
drawing towards
towards dawn,
dawn, and
and you
you have
have slept
slept now
now since
since the
the night’s
night’s beginning.
beginning. Tomorrow,
Tomorrow, perhaps,


of their
their times
times together,
together, of
of course,
course, and
and Bilbo
Bilbo asked
asked how
how things
things were
were going
going in
in the
the lands
lands of
of the
the Mountain.
Mountain. It
It seemed
seemed they
they were
were going
going very
very well.
well. Bard
Bard had
had rebuilt
rebuilt the
the town
town in
in Dale
Dale and
and men
men had
had gathered
gathered to
to him
him from
from the
the Lake
Lake and
and from
from South
South and
and West,
West, and
and all
all the
the valley
valley had
had become
become tilled
tilled again
again and
and rich,
rich, and
and the
the desolation
desolation was
was now
now filled
filled with
with birds
birds and
and blossoms
blossoms in
in spring
spring and
and fruit
fruit and
and feasting
feasting in
in autumn.
autumn. And
And Lake-town
Lake-town was
was refounded
refounded and
and was
was more
more prosperous
prosperous than
than ever,
ever, and
and much
much wealth
wealth went
went up
up and
and down
down the
the Running
R

But remember we don't want to deal with words as strings. We need the indexes. And we need the words we decided to skip (the singletons) to be replaced with the index for UNK.

In [199]:
bigrams = [ (word2index[words[i-1]], word2index[words[i]]) for i in range(1, len(words)) ]

In [200]:
bigramFreqs = Counter(bigrams)

In [201]:
len(bigrams)

95939

In [146]:
bigramFreqs.items()

dict_items([((0, 0), 651), ((0, 1), 2), ((1, 0), 4), ((0, 2), 4), ((2, 3), 1), ((3, 0), 33), ((0, 4), 28), ((4, 5), 7), ((5, 6), 5), ((6, 7), 6), ((7, 8), 527), ((8, 9), 20), ((9, 10), 1), ((10, 11), 2), ((11, 5), 2), ((5, 12), 1), ((12, 13), 1), ((13, 5), 3), ((5, 14), 1), ((14, 0), 1), ((0, 15), 1), ((15, 16), 1), ((16, 17), 1), ((17, 18), 26), ((18, 8), 88), ((8, 0), 717), ((0, 19), 241), ((19, 0), 222), ((0, 20), 850), ((20, 21), 8), ((21, 0), 30), ((0, 22), 1), ((22, 23), 1), ((23, 24), 1), ((24, 5), 1), ((5, 25), 1), ((25, 0), 1), ((0, 6), 2), ((6, 18), 1), ((18, 26), 2), ((26, 7), 2), ((7, 27), 3), ((27, 28), 16), ((28, 29), 8), ((29, 30), 4), ((30, 31), 12), ((31, 32), 1), ((32, 28), 16), ((28, 0), 172), ((0, 27), 48), ((27, 33), 159), ((33, 5), 127), ((5, 34), 1), ((34, 20), 2), ((20, 35), 40), ((35, 36), 1), ((36, 0), 1), ((0, 37), 56), ((37, 38), 8), ((38, 5), 49), ((5, 39), 2), ((39, 40), 1), ((40, 41), 1), ((41, 42), 2), ((42, 5), 56), ((5, 0), 312), ((0, 43), 1), ((43, 44

In [202]:
bigramFreqs[(word2index["of"], word2index["the"])]

809

In [203]:
bigramFreqs[(word2index["of"], word2index["Ralphie"])]

222

In [204]:
bigramFreqs[(word2index["of"],0)]

222

In [205]:
unigrams = [word2index[word] for word in words]

In [111]:
unigrams

[1,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 3,
 4,
 0,
 0,
 0,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 6,
 13,
 14,
 6,
 15,
 0,
 16,
 17,
 18,
 19,
 9,
 0,
 20,
 0,
 21,
 22,
 0,
 23,
 24,
 25,
 6,
 26,
 0,
 0,
 7,
 19,
 27,
 8,
 28,
 29,
 30,
 31,
 32,
 33,
 29,
 0,
 28,
 34,
 6,
 35,
 21,
 36,
 37,
 0,
 38,
 39,
 6,
 40,
 41,
 42,
 43,
 6,
 0,
 44,
 45,
 19,
 6,
 0,
 46,
 47,
 0,
 8,
 9,
 48,
 49,
 50,
 42,
 51,
 32,
 29,
 6,
 0,
 52,
 43,
 6,
 0,
 6,
 53,
 54,
 55,
 56,
 57,
 19,
 0,
 58,
 21,
 0,
 0,
 21,
 0,
 59,
 19,
 0,
 0,
 21,
 60,
 21,
 60,
 20,
 0,
 61,
 0,
 21,
 62,
 63,
 9,
 64,
 34,
 65,
 20,
 0,
 50,
 55,
 66,
 32,
 21,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 9,
 75,
 20,
 9,
 76,
 63,
 50,
 77,
 78,
 79,
 9,
 80,
 61,
 81,
 82,
 41,
 83,
 28,
 63,
 21,
 81,
 84,
 41,
 85,
 51,
 86,
 20,
 87,
 88,
 32,
 89,
 75,
 21,
 90,
 32,
 91,
 92,
 68,
 0,
 61,
 9,
 0,
 0,
 0,
 93,
 0,
 0,
 20,
 0,
 0,
 94,
 39,
 95,
 96,
 0,
 29,
 0,
 0,
 0,
 79,
 97,
 32,
 9,
 98,
 99,
 21,
 100,
 32,
 9,

In [206]:
unigramFreqs = Counter(unigrams)

In [207]:
def unigramProb(unigram): 
    return unigramFreqs[unigram]/(sum(unigramFreqs.values()))

In [208]:
word2index["Gandalf"]

363

In [209]:
unigramProb(363)

0.0010318949343339587

In [210]:
unigramFreqs[363]

99

In [211]:
test1 = "In a hole in the ground there lived a hobbit"

In [213]:
def unigramSentenceProb(words): 
    return np.prod([unigramProb(word2index[word]) for word in words])


In [214]:
unigramSentenceProb(test1.split())

7.5075513010812298e-27

In [215]:
test1.split()

['In', 'a', 'hole', 'in', 'the', 'ground', 'there', 'lived', 'a', 'hobbit']

In [216]:
[word2index[word] for word in test1.split()]

[5, 6, 7, 8, 9, 10, 11, 12, 6, 64]

For a bigram model we need the product of the bigram probs in a sentence. For bigram probs we need the bigram count divided by the unigram count of the first word.

In [217]:
def bigramProb(bigram):
    return (bigramFreqs[bigram]/unigramFreqs[bigram[0]])

In [218]:
def bigramSentenceProb(words):
    return np.prod([bigramProb((word2index[words[i-1]], word2index[words[i]])) for i in range(1, len(words))])

In [219]:
bigramProb((word2index["of"], word2index["the"]))

0.3293973941368078

In [220]:
bigramSentenceProb(test1.split())

5.4976256439726721e-15